subsys/fs/ext2/ext2_diskops.c

```diff
@@ -0,0 +1,1164 @@
+
+	if (itable_offset < 0) {
+		return itable_offset;
+	}
+
+	uint8_t *bytes = (uint8_t *)&BGROUP_INODE_TABLE(&fs->bgroup)[itable_offset];
+
+	for (int i = 0; i < sizeof(struct ext2_disk_inode); ++i) {
+		if (bytes[i] != 0) {
+			return -EINVAL;
+		}
+	}
+	return 0;
+}
+
+int32_t ext2_alloc_inode(struct ext2_data *fs)
+{
+	int rc, r;
+	uint32_t group = 0, set;
+	int32_t global_idx;
+
+	rc = ext2_fetch_block_group(fs, group);
+
+	while (fs->bgroup.bg_free_inodes_count == 0 && rc >= 0) {
+		group++;
+		rc = ext2_fetch_block_group(fs, group);
+		if (rc == -ERANGE) {
+			/* reached last group */
+			return -ENOSPC;
+		}
+	}
+
+	if (rc < 0) {
+		return rc;
+	}
+
+	LOG_DBG("Free inodes (bg): %d", fs->bgroup.bg_free_inodes_count);
+	LOG_DBG("Free inodes (sb): %d", fs->sblock.s_free_inodes_count);
+
+	rc = ext2_fetch_bg_ibitmap(&fs->bgroup);
+	if (rc < 0) {
+		return rc;
+	}
+
+	r = ext2_bitmap_find_free(BGROUP_INODE_BITMAP(&fs->bgroup), fs->block_size);
+	if (r < 0) {
+		LOG_DBG("Cannot find free inode in group %d (rc: %d)", group, r);
+		return r;
+	}
+
+	/* Add 1 because inodes are counted from 1 not 0. */
+	global_idx = group * fs->sblock.s_inodes_per_group + r + 1;
+
+	/* Inode table entry for found inode must be cleared. */
+	if (check_zero_inode(fs, global_idx) != 0) {
+		error_behavior(fs,  "Inode is not cleared in inode table!");
+		return -EINVAL;
+	}
+
+	LOG_DBG("Found free inode %d in group %d (global_idx: %d)", r, group, global_idx);
+
+	rc = ext2_bitmap_set(BGROUP_INODE_BITMAP(&fs->bgroup), r, fs->block_size);
+	if (rc < 0) {
+		return rc;
+	}
+
+	fs->bgroup.bg_free_inodes_count -= 1;
+	fs->sblock.s_free_inodes_count -= 1;
+
+	set = ext2_bitmap_count_set(BGROUP_INODE_BITMAP(&fs->bgroup), fs->sblock.s_inodes_count);
+
+	if (set != fs->sblock.s_inodes_count - fs->sblock.s_free_inodes_count) {
+		error_behavior(fs, "Wrong number of used inodes in superblock and bitmap");
+		return -EINVAL;
+	}
+
+	rc = ext2_commit_superblock(fs);
+	if (rc < 0) {
+		LOG_DBG("super block write returned: %d", rc);
+		return -EIO;
+	}
+	rc = ext2_commit_bg(fs);
+	if (rc < 0) {
+		LOG_DBG("block group write returned: %d", rc);
+		return -EIO;
+	}
+	rc = ext2_write_block(fs, fs->bgroup.inode_bitmap);
+	if (rc < 0) {
+		LOG_DBG("block bitmap write returned: %d", rc);
+		return -EIO;
+	}
+
+	LOG_DBG("Free inodes (bg): %d", fs->bgroup.bg_free_inodes_count);
+	LOG_DBG("Free inodes (sb): %d", fs->sblock.s_free_inodes_count);
+
+	return global_idx;
+}
+
+int ext2_free_block(struct ext2_data *fs, uint32_t block)
+{
+	LOG_DBG("Free block %d", block);
+
+	/* Block bitmaps tracks blocks starting from s_first_data_block. */
+	block -= fs->sblock.s_first_data_block;
+
+	int rc;
+	uint32_t group = block / fs->sblock.s_blocks_per_group;
+	uint32_t off = block % fs->sblock.s_blocks_per_group;
+	uint32_t set;
+
+	rc = ext2_fetch_block_group(fs, group);
+	if (rc < 0) {
+		return rc;
+	}
+
+	rc = ext2_fetch_bg_bbitmap(&fs->bgroup);
+	if (rc < 0) {
+		return rc;
+	}
```

Comment:
If `off` is incorrect, which will be checked by `ext2_bitmap_unset`, then this seem pointless.

Reply:
https://github.com/zephyrproject-rtos/zephyr/pull/55152#discussion_r1268024770 also applies here